app/assets/javascripts/components/story/CollapsedStory/index.js

```diff
@@ -25,65 +29,187 @@ const storyClassName = (story, additionalClassname = '') => {
       'Story--estimated': !isStoryNotEstimated
     },
     additionalClassname
-  );
-};
+  )
+}
 
-export const CollapsedStory = ({ 
-  onToggle, 
-  story, 
-  updateCollapsedStory, 
-  project, 
-  className, 
-  title, 
-  highlight, 
-  isHighlightable 
-}) =>
-  <div
-    className={storyClassName(story, className)}
-    onClick={onToggle}
-    title={title}
-  >
-    <StoryPopover story={story}>
-      <div className='Story__icons-block'>
-        <StoryIcon storyType={story.storyType} />
-        <CollapsedStoryEstimate estimate={story.estimate} />
-        <StoryDescriptionIcon description={story.description} />
-      </div>
-    </StoryPopover>
-
-    <CollapsedStoryInfo story={story} />
-
-    <CollapsedStoryStateActions story={story}
-      onUpdate={(newAttributes) =>
-        updateCollapsedStory(story.id, project.id, newAttributes)}
-    />
-    {
-      isHighlightable && 
-      <CollapsedStoryFocusButon onClick={() => highlight(story.id)} />
+const type = {
+  story: 'STORY'
+}
+
+const calculatePosition = (aboveStory, bellowStory, storiesArray, index) => {
+  if (bellowStory === undefined) {
+    return (Number(storiesArray[index].position) + 1)
+  } else if (aboveStory === undefined) {
+    return (Number(storiesArray[index].position) - 1);
+  }
+  return (Number(bellowStory.position) + Number(aboveStory.position)) / 2;
+}
+
+const getNewPosition = (item, story, storiesArray, index) => {
+  if(item.index > index) {
+    return calculatePosition(storiesArray[index - 1], story, storiesArray, index);
+  }
+  return calculatePosition(story, storiesArray[index + 1], storiesArray, index);
+}
+
+const isDraggedInTheSamePlaceOfHolder = (dragIndex, hoverIndex) => dragIndex === hoverIndex
+const isDraggedBellowTheHover = (dragIndex, hoverIndex, hoverClientY, hoverMiddleY) => dragIndex < hoverIndex && hoverClientY < hoverMiddleY
+const isDraggedAboveTheHover = (dragIndex, hoverIndex, hoverClientY, hoverMiddleY) => dragIndex > hoverIndex && hoverClientY > hoverMiddleY
+const canContinue = (dragIndex, hoverIndex, hoverClientY, hoverMiddleY) =>
+isDraggedBellowTheHover(dragIndex, hoverIndex, hoverClientY, hoverMiddleY) || isDraggedAboveTheHover(dragIndex, hoverIndex, hoverClientY, hoverMiddleY)
+
+const checkStoryPosition = (ref, dragIndex, hoverIndex, monitor) => {
+  const hoverBoundingRect = ref.current.getBoundingClientRect()
+  const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2
+  const clientOffset = monitor.getClientOffset()
+  const hoverClientY = clientOffset.y - hoverBoundingRect.top
+
+  return isDraggedInTheSamePlaceOfHolder(dragIndex, hoverIndex)
+    || canContinue(dragIndex, hoverIndex, hoverClientY, hoverMiddleY)
+}
+
+const canChangePosition = (ref, monitor, storySource, hoverIndex, newPosition) => {
+  if(!ref.current) {
```

Comment:
So, if any os theses conditions are `true` you wanna return `true`, right?
So you can rewrite it:
```js
const canChangePosition = (ref, monitor, storySource, hoverIndex, newPosition) => 
  !ref.current ||
    storySource.position === newPosition || 
    storySource.index === hoverIndex || 
    checkStoryPosition(ref, storySource.index, hoverIndex, monitor)
```

Reply:
Like that.